### PR TITLE
feat: add Coordinator PR review assignment

### LINKED ISSUE #411: feat: add Coordinator PR review assignment
## Goal

Let Coordinator decide whether a PR should receive Looper review, then request an eligible reviewer in local-only and Routed modes.

## Scope

- Add Coordinator PR review-assignment policy.
- In local-only mode, Coordinator may request review from the configured local Reviewer GitHub identity.
- In Routed mode, only the lease-holder Coordinator Node may perform review-assignment GitHub mutations.
- Routed PR assignment chooses an eligible Reviewer Node, applies coarse GitHub authority, then adds exact target label last:
  1. revalidate lease;
  2. request review from selected Reviewer GitHub identity;
  3. revalidate lease;
  4. add/repair exactly one `looper:target:<node_name>` label last.
- Candidate Nodes come from Network membership and project eligibility.
- Existing Reviewer filters and self-review rules must be considered before selecting a target.
- If no eligible Reviewer Node exists, record/status `no-eligible-node` and do not request review or add target label.

## Authority constraints

- Coordinator PR assignment decision, or an existing human/external GitHub review request, is the review work-intent authority.
- Reviewer still claims only from current GitHub review-request state.
- `loopernet` does not mutate GitHub.
- Audit does not override current human changes.

## Acceptance criteria

- New/admissible PRs can receive a review request in local-only mode.
- New/admissible PRs can receive review request + exact target label in Routed mode.
- Routed mutations require valid Coordinator lease revalidation before each side-effect boundary.
- Target label is written last to avoid premature claim of partially assigned PRs.
- Selection avoids permanently unclaimable targets by applying existing Reviewer filters before target selection.
- Existing human/external review requests remain valid Reviewer work intent.
- If a human removes the review request, later behaviour is based on fresh Coordinator policy/current GitHub state, not stale audit.
- Tests cover local-only assignment, Routed assignment, duplicate GitHub identity, self-review/filter exclusion, lease loss mid-sequence, and no eligible Node.

### PR DESCRIPTION
**Pull request by mrcfps**
## Summary
- add Coordinator-driven PR review assignment for local-only and Routed projects
- enforce routed lease revalidation and exact target-label sequencing around review-assignment mutations
- publish reviewer project capabilities over loopernet and cover local/routed assignment edge cases in tests

## Validation
- go test ./...
- go vet ./...
- go build ./...

Closes #411

### CODE REVIEW

File: internal/coordinator/runner.go [hunk 1]
```diff
@@ -902,6 +914,283 @@ func normalizeLogin(login string) string {
...
+
+func (r *Runner) applyReviewAssignments(ctx context.Context, projectID, repo, cwd string) error {
+	if r == nil || r.github == nil || r.config == nil {
+		return nil
+	}
+	roles := config.ProjectRoleConfigs(*r.config, projectID)
+	trigger := roles.Reviewer.Discovery.Triggers
+	policy := networkpolicy.ProjectPolicyForProject(*r.config, projectID)
+	if !roles.Reviewer.Discovery.AutoDiscovery {
+		return nil
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/coordinator/runner.go]: @chatgpt-codex-connector This has been fixed in 6eb2ffa.

Updated internal/coordinator/runner.go so routed review assignment still runs when local reviewer autodiscovery is disabled and routed PR authority comes from reviewer capabilities; added regression coverage in internal/coordinator/runner_actions_test.go and verify fake-gh allowlists.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by mrcfps** [comment]
**Looper fixer round complete** — 6eb2ffa

- ✅ Review comment on `internal/coordinator/runner.go` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/440#discussion_r3278734587)
  - Updated internal/coordinator/runner.go so routed review assignment still runs when local reviewer autodiscovery is disabled and routed PR authority comes from reviewer capabilities; added regression coverage in internal/coordinator/runner_actions_test.go and verify fake-gh allowlists.
- • Review comment on `internal/coordinator/runner.go` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/440#discussion_r3278778185)
  - Agent did not provide a decision for this thread


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

File: internal/coordinator/runner.go [hunk 2]
```diff
@@ -1128,6 +1131,283 @@ func normalizeLogin(login string) string {
...
+		detail, err := r.github.ViewPullRequest(ctx, githubinfra.ViewPullRequestInput{Repo: repo, PRNumber: prNumber, CWD: cwd})
+		if err != nil {
+			return err
+		}
+		if !strings.EqualFold(strings.TrimSpace(detail.State), "open") {
+			continue
+		}
+		authority := labelsMatch(detail.Labels, trigger.Labels, trigger.LabelMode) || len(detail.ReviewRequests) > 0 || len(detail.ReviewRequestUsers) > 0
+		if !authority {
+			continue
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/coordinator/runner.go]: @chatgpt-codex-connector I'm not making a code change for this thread.

I could not safely verify current PR metadata because the required gh pr view call was rate-limited, so I did not make or validate any branch changes for this thread.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

File: internal/coordinator/runner.go [hunk 3]
```diff
@@ -1128,6 +1131,320 @@ func normalizeLogin(login string) string {
...
+			return err
+		}
+		if !strings.EqualFold(strings.TrimSpace(detail.State), "open") {
+			continue
+		}
+		authority := len(detail.ReviewRequests) > 0 || len(detail.ReviewRequestUsers) > 0
+		if routed {
+			authority = authority || routedReviewAssignmentAuthority(projectID, routedMemberships, detail)
+		} else {
+			authority = authority || labelsMatch(detail.Labels, trigger.Labels, trigger.LabelMode)
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/coordinator/runner.go]: @chatgpt-codex-connector I'm not making a code change for this thread.

No new change was needed for this thread because the current branch already gates local assignment through reviewAssignmentMatchesTrigger in internal/coordinator/runner.go, and TestRunnerSkipsLocalReviewerAssignmentWithoutReviewRequestWhenRequired already covers the requireReviewRequest case.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

File: internal/coordinator/runner.go [hunk 4]
```diff
@@ -1128,6 +1131,320 @@ func normalizeLogin(login string) string {
...
+	}
+	return "login:" + normalizeLogin(login)
+}
+
+func leaseProbeURL(repo string, prNumber int64) string {
+	parts := strings.Split(strings.TrimSpace(repo), "/")
+	if len(parts) >= 3 && strings.Contains(parts[0], ".") {
+		return fmt.Sprintf("https://%s/%s/%s/pull/%d", parts[0], parts[1], parts[2], prNumber)
+	}
+	return fmt.Sprintf("https://github.com/%s/pull/%d", strings.TrimSpace(repo), prNumber)
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/coordinator/runner.go]: @chatgpt-codex-connector I'm not making a code change for this thread.

No new change was needed for this thread because leaseProbeURL in internal/coordinator/runner.go already preserves single-label hosts on this branch, and TestLeaseProbeURLPreservesSingleLabelHost already covers that regression.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

File: internal/coordinator/runner.go [hunk 5]
```diff
@@ -1128,6 +1131,327 @@ func normalizeLogin(login string) string {
...
+		}
+		if reviewAssignmentMatchesTrigger(detail, reviewerTriggerFromCapability(projectCapability)) {
+			return true
+		}
+	}
+	return false
+}
+
+func reviewerTriggerFromCapability(capability protocol.ReviewerProjectCapability) config.ReviewerRoleTriggersConfig {
+	return config.ReviewerRoleTriggersConfig{IncludeDrafts: capability.IncludeDrafts, EnableSelfReview: capability.EnableSelfReview, Labels: append([]string(nil), capability.Labels...), LabelMode: config.LabelMode(capability.LabelMode)}
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/coordinator/runner.go]: @chatgpt-codex-connector This has been fixed in 9d22844.

I updated routed reviewer capability propagation to carry requireReviewRequest through internal/network/protocol/protocol.go, internal/network/client/manager.go, and internal/coordinator/runner.go, defaulting missing legacy payloads to require review requests. I also added regression coverage in internal/coordinator/runner_actions_test.go and internal/network/client/manager_test.go.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by mrcfps** [comment]
**Looper fixer round complete** — 9d22844

- ⏸️ Review comment on `internal/coordinator/runner.go` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/440#discussion_r3278819103)
  - No new change was needed for this thread because the current branch already gates local assignment through reviewAssignmentMatchesTrigger in internal/coordinator/runner.go, and TestRunnerSkipsLocalReviewerAssignmentWithoutReviewRequestWhenRequired already covers the requireReviewRequest case.
- ⏸️ Review comment on `internal/coordinator/runner.go` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/440#discussion_r3278819105)
  - No new change was needed for this thread because leaseProbeURL in internal/coordinator/runner.go already preserves single-label hosts on this branch, and TestLeaseProbeURLPreservesSingleLabelHost already covers that regression.
- ✅ Review comment on `internal/coordinator/runner.go` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/440#discussion_r3278868775)
  - I updated routed reviewer capability propagation to carry requireReviewRequest through internal/network/protocol/protocol.go, internal/network/client/manager.go, and internal/coordinator/runner.go, defaulting missing legacy payloads to require review requests. I also added regression coverage in internal/coordinator/runner_actions_test.go and internal/network/client/manager_test.go.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>
